nameko/messaging.py

```diff
@@ -166,26 +182,15 @@ def retry_policy(self):
         return DEFAULT_RETRY_POLICY
 
     def setup(self):
-
-        exchange = self.exchange
-        queue = self.queue
+        if self.exchange is not None:
```

Comment:
In the `EventDispatcher` case the exchange is derived from the service name, which is only available after the DependencyProvider is bound to a container. We know that has happened by the time `setup` executes.

I may try adding a conditional to the constructor to see if it makes things clearer